service/worker/scheduler/workflow.go

```diff
@@ -0,0 +1,800 @@
+		ScheduleTime: start.ActualTime,
+		ActualTime:   res.RealStartTime,
+		StartWorkflowResult: &commonpb.WorkflowExecution{
+			WorkflowId: workflowID,
+			RunId:      res.RunId,
+		},
+	}, nil
+}
+
+func (s *scheduler) identity() string {
+	return fmt.Sprintf("temporal-scheduler-%s-%s", s.State.Namespace, s.State.ScheduleId)
+}
+
+func (s *scheduler) addSearchAttrs(
+	attrs *commonpb.SearchAttributes,
+	nominal time.Time,
+) *commonpb.SearchAttributes {
+	fields := maps.Clone(attrs.GetIndexedFields())
+	if p, err := payload.Encode(nominal); err == nil {
+		fields[searchAttrStartTime] = p
+	}
+	if p, err := payload.Encode(s.State.ScheduleId); err == nil {
+		fields[searchAttrScheduleById] = p
+	}
+	return &commonpb.SearchAttributes{
+		IndexedFields: fields,
+	}
+}
+
+func (s *scheduler) refreshWorkflows(executions []*commonpb.WorkflowExecution) {
+	ctx := workflow.WithLocalActivityOptions(s.ctx, defaultLocalActivityOptions)
+	futures := make([]workflow.Future, len(executions))
+	for i, ex := range executions {
+		req := &schedspb.WatchWorkflowRequest{
+			Namespace:   s.State.Namespace,
+			NamespaceId: s.State.NamespaceId,
+			// Note: do not send runid here so that we always get the latest one
+			Execution:           &commonpb.WorkflowExecution{WorkflowId: ex.WorkflowId},
+			FirstExecutionRunId: ex.RunId,
+			LongPoll:            false,
+		}
+		futures[i] = workflow.ExecuteLocalActivity(ctx, s.a.WatchWorkflow, req)
+	}
+	for i, ex := range executions {
+		s.processWatcherResult(ex.WorkflowId, futures[i])
+	}
+}
+
+func (s *scheduler) startLongPollWatcher(ex *commonpb.WorkflowExecution) {
+	if s.watchingFuture != nil {
+		s.logger.Error("startLongPollWatcher called with watcher already running")
+		return
+	}
+
+	ctx := workflow.WithActivityOptions(s.ctx, workflow.ActivityOptions{
+		StartToCloseTimeout: 365 * 24 * time.Hour,
+		RetryPolicy: &temporal.RetryPolicy{
+			InitialInterval: 1 * time.Second,
+			MaximumInterval: 30 * time.Second,
+		},
+		HeartbeatTimeout: 65 * time.Second,
+	})
+	req := &schedspb.WatchWorkflowRequest{
+		Namespace:   s.State.Namespace,
+		NamespaceId: s.State.NamespaceId,
+		// Note: do not send runid here so that we always get the latest one
+		Execution:           &commonpb.WorkflowExecution{WorkflowId: ex.WorkflowId},
+		FirstExecutionRunId: ex.RunId,
+		LongPoll:            true,
+	}
+	s.watchingFuture = workflow.ExecuteActivity(ctx, s.a.WatchWorkflow, req)
+	s.watchingWorkflowId = ex.WorkflowId
+}
+
+func (s *scheduler) cancelWorkflow(ex *commonpb.WorkflowExecution) {
+	ctx := workflow.WithLocalActivityOptions(s.ctx, defaultLocalActivityOptions)
+	areq := &schedspb.CancelWorkflowRequest{
+		NamespaceId: s.State.NamespaceId,
+		Namespace:   s.State.Namespace,
+		RequestId:   s.newUUIDString(),
+		Identity:    s.identity(),
+		Execution:   ex,
+		Reason:      "cancelled by schedule overlap policy",
+	}
+	workflow.ExecuteLocalActivity(ctx, s.a.CancelWorkflow, areq)
+	// do not wait for cancel to complete
```

Comment:
i think we have to wait for the RequestCancelWorkflowExecution API call to return. If we don't wait here, the local activity might not get a chance to run and thus the API call might not be made.

Reply:
will do (and will set retry policy)